### PR TITLE
A bunch of changes

### PR DESCRIPTION
**Pull request by miklos1**
This pull request brings a bunch of changes which somewhat build upon each other:
 * Docstrings for impero.py
 * Pull COFFEE-independent code out of coffee.py into impero_utils.py
 * A sensible DAG-visiting infrastructure
 * Performance improvements for ComponentTensor inlining
 * Performance improvements for IndexSum expansion
 * Docstrings for the optimisation module and the newly added visitor tools

I would like to emphasise that these changes **improve compilation speed** significantly for complicated forms, especially with the default settings (IndexSum expansion for small loops).

### CODE REVIEW

File: tsfc/node.py [hunk 1]
```diff
@@ -107,3 +107,94 @@ def traversal(expression_dags):
...
+    """Reuse if untouched recipe"""
+    new_children = map(self, node.children)
+    if all(nc == c for nc, c in zip(new_children, node.children)):
+        return node
+    else:
+        return node.reconstruct(*new_children)
+
+
+def reuse_if_untouched_arg(node, self, arg):
+    """Reuse if touched recipe propagating and extra argument"""
```
Comment (wence-) [inline tsfc/node.py]: s/and/an/?

File: tsfc/optimise.py [hunk 2]
```diff
@@ -2,89 +2,103 @@
...
-            for i in xrange(node.index.extent):
-                result = Sum(result, Indexed(ct, (i,)))
-            return result
-        else:
-            return node.reconstruct(*[cached_handle(child) for child in node.children])
+            return Indexed(new_child, multiindex)
+
+
+def filtered_replace_indices(node, self, subst):
+    """Wrapper for :func:`replace_index`.  At each call removes
```
Comment (wence-) [inline tsfc/optimise.py]: `replace_indices`?

File: tsfc/optimise.py [hunk 3]
```diff
@@ -2,89 +2,103 @@
...
+    substitution rules that do not apply."""
+    filtered_subst = tuple((k, v) for k, v in subst if k in node.free_indices)
+    return replace_indices(node, self, filtered_subst)
+
+
+def filtered_replace_indices_top(node, self, subst):
+    """Wrapper for :func:`replace_index`.  At each call removes
```
Comment (wence-) [inline tsfc/optimise.py]: ditto.

File: tsfc/optimise.py (hunk 2 above)
Comment (miklos1) [inline tsfc/optimise.py]: Thanks. I'll fix.

File: tsfc/node.py (hunk 1 above)
Comment (miklos1) [inline tsfc/node.py]: Yes.

File: tsfc/node.py [hunk 4]
```diff
@@ -107,3 +107,94 @@ def traversal(expression_dags):
...
+        self.function = function
+
+    def __call__(self, node, arg):
+        cache_key = (node, arg)
+        try:
+            return self.cache[cache_key]
+        except KeyError:
+            result = self.function(node, self, arg)
+            self.cache[cache_key] = result
+            return result
```
Comment (wence-) [inline tsfc/node.py]: Could these implementations be merged by just allowing `*args`?

Comment (miklos1) [inline tsfc/node.py]: Yes, but that makes processing a bit slower. I would keep them for now.

**Pull request merged**